### PR TITLE
Support zero-byte read in gRPC client

### PR DESCRIPTION
**Pull request by JamesNK**
Avoid renting and using a buffer until a response is available using a zero-byte read of the response.

### CODE REVIEW

File: src/Grpc.Net.Client/Internal/StreamExtensions.cs [hunk 1]
```diff
@@ -62,6 +62,13 @@ private static Status CreateUnknownMessageEncodingMessageStatus(string unsupport
             GrpcCallLog.ReadingMessage(call.Logger);
             cancellationToken.ThrowIfCancellationRequested();
 
+#if NET7_0_OR_GREATER
+            // Start with zero-byte read.
+            // A zero-byte read avoids renting buffer until the response is ready. Especially useful for long running streaming calls.
+            var readCount = await responseStream.ReadAsync(Memory<byte>.Empty, cancellationToken).ConfigureAwait(false);
```
Comment (MihaZupan) [inline src/Grpc.Net.Client/Internal/StreamExtensions.cs:68]: Is the common case with gRPC that the whole message will be received quickly once you start receiving data?
That is, the server won't send you 10 bytes, wait for a second, then send you the rest?

Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/StreamExtensions.cs:68]: There can be delays between messages when streaming, but a message is flushed by the server to the client all at once.

File: src/Grpc.Net.Client/Internal/StreamExtensions.cs [hunk 2]
```diff
@@ -62,6 +62,13 @@ private static Status CreateUnknownMessageEncodingMessageStatus(string unsupport
             GrpcCallLog.ReadingMessage(call.Logger);
             cancellationToken.ThrowIfCancellationRequested();
 
+#if NET7_0_OR_GREATER
```
Comment (davidfowl) [inline src/Grpc.Net.Client/Internal/StreamExtensions.cs:65]: Why the ifdef? Something to do with the underlying stream impl?

Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/StreamExtensions.cs:65]: Yes. Some HttpHandler stream implementations in old frameworks don't like zero byte buffers, e.g. `Microsoft.AspNetCore.TestHost.ResponseBodyReaderStream`

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 3]
```diff
@@ -56,6 +56,17 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
         switch (_state)
         {
             case ResponseState.Ready:
+                if (data.Length == 0)
+                {
+                    // Handle zero byte reads.
+                    return 0;
```
Comment (halter73) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:62]: Shouldn't we try a zero-length read on `_inner` before immediately returning 0? It looks like this logic has the risk stalling in `TryReadDataAsync` during the next call to `ReadAsync` which will result in a 4 KB buffer or larger being rented even though it will be sliced to only 5 bytes for the header.

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 4]
```diff
@@ -88,7 +99,7 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
                 }
 
                 _contentRemaining = length;
-                // If there is no content then state is still ready
+                // If there is no content then state is reset to ready.
                 _state = _contentRemaining > 0 ? ResponseState.Content : ResponseState.Ready;
                 return 5;
```
Comment (halter73) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:104]: I know this isn't new, but this seems really weird to me. Especially given the above `// Should never get here. Client always passes 5 to read the header` comment. If this is intended to be consumed in a special way and not as an abstract Stream where I can pass any buffer length, `GrpcWebResponseStream` should not be implementing `Stream`.

Instead, I think it would be better to implement `ReadMessageAsync` directly here in a non-Stream type rather than as an extension method on `StreamExtensions`. That way you don't end up having to do things like parse the header twice. The extension method does seem resilient enough to handle the inner stream returning a partial header just like `TryReadDataAsync` in this `GrpcWebResponseStream` does, so there seems to be a lot of redundant logic.

Comment (JamesNK) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs]: The stream is combined with a custom `HttpContent`, which intercepts the incoming HTTP response body and transforms it. Realistically it will only ever be used by Grpc.Net.Client, which always passes a buffer larger than 5 bytes (unless a zero byte read).

But you're right that it doesn't completely conform to the `Stream` spec. I've updated it to accept any size buffer.

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 5]
```diff
@@ -52,59 +58,113 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
...
-                else
+            case ResponseState.Header:
                 {
-                    // Should never get here. Client always passes 5 to read the header.
-                    throw new InvalidOperationException("Buffer is not large enough for header");
+                    headerBuffer = data.Length >= _contentRemaining ? data.Slice(0, _contentRemaining) : data;
+                    var success = await TryReadDataAsync(_inner, headerBuffer, cancellationToken).ConfigureAwait(false);
+                    if (!success || headerBuffer.Length == 0)
+                    {
+                        return 0;
```
Comment (halter73) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:90]: Can `_contentRemaining` ever be zero here? If so, this seems like it could incorrectly report the end of stream. It looks like this might be impossible because you should transition to the `Ready` state first, so maybe just add an assert?

```suggestion
                        Debug.Assert(_contentRemaining > 0);
                        return 0;
```

**Review by halter73 — APPROVED** [review]

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs (hunk 5 above)
Comment (halter73) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:90]: Also, doesn't `TryReadDataAsync` always return false given a zero-length header buffer?

```suggestion
                    Debug.Assert(_contentRemaining > 0);
                    if (!success)
                    {
                        return 0;
```

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 6]
```diff
@@ -52,59 +58,115 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
 #if NETSTANDARD2_0
         var data = buffer.AsMemory(offset, count);
 #endif
+        Memory<byte> headerBuffer = Memory<byte>.Empty;
```
Comment (JamesNK) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:61]:
```suggestion
        var headerBuffer = Memory<byte>.Empty;
```

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 7]
```diff
@@ -31,11 +33,15 @@ internal class GrpcWebResponseStream : Stream
 {
     // This uses C# compiler's ability to refer to static data directly. For more information see https://vcsjones.dev/2019/02/01/csharp-readonly-span-bytes-static
     private static ReadOnlySpan<byte> BytesNewLine => new byte[] { (byte)'\r', (byte)'\n' };
```
Comment (mgravell) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:35]: drive-by observation: `"\r\n"u8;` (if you're happy re lang-ver)

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 8]
```diff
@@ -52,59 +58,115 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
...
+                        }
+                    }
+
+                    var read = headerBuffer.Length;
+
+                    // The buffer was less than header length either because this is the first read and the passed in buffer is small,
+                    // or it is an additonal read to finish getting header data.
+                    if (headerBuffer.Length < HeaderLength)
+                    {
+                        _headerBuffer ??= new byte[HeaderLength];
```
Comment (mgravell) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:113]: this `byte[]` troubles me; it is always `5`, right? well, we can get that far more cheaply by abusing a `long` field (which is the same size as the `byte[]` *reference*, but without needing the `byte[]`) - perhaps thunking as a `Span<byte>`:

``` c#
    private long _header;
    private Span<byte> GetHeader() => MemoryMarshal.Cast<long, byte>(new Span<long>(ref _header)).Slice(0, 5);
```

similar can be done on down-level using `unsafe`, or you can just use shifts

Is it worth the evil, though? perhaps:

``` c#
#if NET7_0_OR_GREATER
    private long _header;
    private Span<byte> GetHeader() => MemoryMarshal.Cast<long, byte>(new Span<long>(ref _header)).Slice(0, 5);
#else
    private byte[]? _header;
    private Span<byte> GetHeader() => _header ??= new byte[5];
#endif
```

Comment (MihaZupan) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:113]:
```c#
MemoryMarshal.CreateSpan(ref Unsafe.As<long, byte>(ref _headerBuffer), 5);
```
will likely get better codegen

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 9]
```diff
@@ -52,59 +58,115 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
...
-
-                var isTrailer = IsBitSet(compressed, pos: 7);
-                if (isTrailer)
+            case ResponseState.Trailer:
                 {
+                    Debug.Assert(headerBuffer.Length > 0);
+
+                    // The trailer needs to be completely read before returning 0 to the caller.
+                    // Ensure buffer is large enough to contain the trailer header.
+                    if (headerBuffer.Length < 5)
```
Comment (mgravell) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:154]: /s/5/HeaderLength ?

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 10]
```diff
@@ -52,59 +58,115 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
...
+                    {
+                        var newBuffer = new byte[HeaderLength];
```
Comment (mgravell) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:156]: as prev

**Review by mgravell — COMMENTED** [review]
(minor comments added)

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs (hunk 8 above)
Comment (mgravell) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:113]: yeah, I thought about that alternative - it does lose the bounds check, but on the other hand, it will never be true that `sizeof(long) < 5`, so that bounds check is redundant... but: what *really* makes that tempting is that `MemoryMarshal.CreateSpan` approach allows us to go all the way back over all netcoreapp2+ and netstandard2.1, so: much better TFM coverage 👍

Comment (JamesNK) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:113]: Header buffer is passed to async methods and needs to be `Memory<byte>`.

This allocation only happens if someone uses a small buffer (less than 5 bytes) with the stream. The primary user of the stream is `Grpc.Net.Client`, which uses a much larger buffer.

Realistically, I doubt people will hit this allocation in the real world.

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs [hunk 11]
```diff
@@ -52,59 +56,113 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
...
+                        if (!success)
+                        {
+                            return 0;
+                        }
+                        headerBuffer = newBuffer;
+                    }
+                    var length = (int)BinaryPrimitives.ReadUInt32BigEndian(headerBuffer.Span.Slice(1));
+
                     await ReadTrailersAsync(length, data, cancellationToken).ConfigureAwait(false);
                     return 0;
```
Comment (amcasey) [inline src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:164]: What's a trailer and why do we always return "0 bytes read" for them?

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs (hunk 8 above)
Comment (amcasey) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:109]: Could we just throw `NotSupportedException` or something if someone tries to read fewer than `HeaderCount` bytes?  This seems like a lot of complexity for code we don't expect or want people to use.

File: test/Grpc.Net.Client.Web.Tests/GrpcWebResponseStreamTests.cs [hunk 12]
```diff
@@ -59,6 +59,110 @@ public async Task ReadAsync_EmptyMessage_ParseMessageAndTrailers()
         Assert.AreEqual("0", trailingHeaders.GetValues("grpc-status").Single());
     }
 
+    [Test]
+    public async Task ReadAsync_HasMessage_OneByteBuffer_ParseMessageAndTrailers()
```
Comment (amcasey) [inline test/Grpc.Net.Client.Web.Tests/GrpcWebResponseStreamTests.cs:63]: Other than the internal state checks, these tests would pass without the product change, correct?

**Review by amcasey — APPROVED** [review]
Seems sensible.  To confirm my understanding of the motivation, this won't _reduce_ renting of buffers but it will _defer_ renting until they're actually needed, reducing contention?

File: src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs (hunk 11 above)
Comment (mgravell) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:164]: My read is that they're not pard of the content payload, and thus we're not reporting them as part of the Stream content; likewise, once we've seen a trailer we know we're at the EOF, hence: return 0 to signal that. But I'm very interested in whether my interpretation was right!

Comment (halter73) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:164]: You're correct. Typically, gRPC uses HTTP/2's built-in trailing header functionality for the "grpc-status" and "grpc-message" trailers, but [gRPC Web encodes this in the response body](https://fuchsia.googlesource.com/third_party/grpc/+/refs/tags/v1.6.5/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2). Waiting for the response body to complete before sending the status allows services to report errors that occur after they start streaming a response.

[HTTP/1.1 also technically supports trailing headers for chunked responses](https://httpwg.org/specs/rfc9112.html#chunked.trailer.section), but support is spotty. I assume that's why gRPC Web doesn't rely on it.

Comment (JamesNK) [reply src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs:164]: > My read is that they're not pard of the content payload, and thus we're not reporting them as part of the Stream content; likewise, once we've seen a trailer we know we're at the EOF, hence: return 0 to signal that. But I'm very interested in whether my interpretation was right!

Correct.

File: test/Shared/SyncPointMemoryStream.cs [hunk 13]
```diff
@@ -37,6 +39,17 @@ public SyncPointMemoryStream(bool runContinuationsAsynchronously = true)
         _awaiter = SyncPoint.Create(out _syncPoint, _runContinuationsAsynchronously);
     }
 
+    /// <summary>
+    /// End stream and wait for at least one more read that returns zero bytes.
+    /// Note that because of zero-byte reads, the stream maybe read multiple times.
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply test/Shared/SyncPointMemoryStream.cs:44]:
```suggestion
    /// Note that because of zero-byte reads, the stream may be read multiple times.
```
